File: src/modules/self_role.ts
import
{
	ApplicationCommandData,
	MessageComponentInteraction,
	CommandInteraction,
	CommandInteractionOption,
	Constants,
	DiscordAPIError,
	GuildMember,
	Interaction,
	MessageActionRowOptions,
	MessageButtonOptions,
	MessageOptions,
	Role,
	TextChannel,
	Snowflake,
} from "discord.js";
import { getConnection } from "typeorm";

import SelfRole from "../entity/SelfRole";
import { IModuleConfig, Module } from "../structures/Module";
import { chunkArray } from "../util";

const { ApplicationCommandOptionTypes, MessageButtonStyles, MessageComponentTypes } = Constants;

const { self_role: { enabled, channel: channelID } } = require("../../config.json");

const CUSTOM_ID_PREFIX = "self-role-";

enum SubCommand
// eslint-disable-next-line @typescript-eslint/indent
{
	add,
	remove,
	show,
	deploy,
	cleanup
}

export const SelfRoleCommand: ApplicationCommandData = {
	name: "selfrole",
	description: "Manage available self roles.",
	defaultPermission: false,
	options: [
		{
			name: SubCommand[SubCommand.add],
			type: ApplicationCommandOptionTypes.SUB_COMMAND,
			description: "Add a role to the list of available self roles.",
			options: [{
				name: "role",
				type: ApplicationCommandOptionTypes.ROLE,
				description: "The role to add to the list of self roles.",
				required: true,
			},
			{
				name: "style",
				type: ApplicationCommandOptionTypes.INTEGER,
				description: "The button style to use",
				choices: [
					{
						name: "Primary (Blurple)",
						value: MessageButtonStyles.PRIMARY,
					},
					{
						name: "Secondary (Grey)",
						value: MessageButtonStyles.SECONDARY,
					},
					{
						name: "Success (Green)",
						value: MessageButtonStyles.SUCCESS,
					},
					{
						name: "Danger (Red)",
						value: MessageButtonStyles.DANGER,
					},
				],
			},
			{
				name: "emoji",
				type: ApplicationCommandOptionTypes.STRING,
				description: "The emoji to use. (Make sure its actually valid)",
			}],
		},
		{
			name: SubCommand[SubCommand.remove],
			type: ApplicationCommandOptionTypes.SUB_COMMAND,
			description: "Remove a role from the list of available self roles.",
			options: [{
				name: "role",
				type: ApplicationCommandOptionTypes.ROLE,
				description: "The role to remove from the list of self roles.",
				required: true,
			}],
		},
		{
			name: SubCommand[SubCommand.show],
			type: ApplicationCommandOptionTypes.SUB_COMMAND,
			description: "Show all available self roles.",
		},
		{
			name: SubCommand[SubCommand.deploy],
			type: ApplicationCommandOptionTypes.SUB_COMMAND,
			description: "Deploys the self role message.",
			options: [{
				name: "message",
				type: ApplicationCommandOptionTypes.STRING,
				description: "The id of the message in the self roles channel to update.",
			}],
		},
		{
			name: SubCommand[SubCommand.cleanup],
			type: ApplicationCommandOptionTypes.SUB_COMMAND,
			description: "Removes deleted roles from the list of self assignable roles.",
		},
	],
};

class SelfRoleModule extends Module
{
	public constructor(config: IModuleConfig)
	{
		super({
			client: config.client,
			enabled,
			eventName: "interaction",
		});
	}

	protected override async handle(interaction: Interaction): Promise<void>
	{
		if (!interaction.guildID) return;
		if (interaction instanceof MessageComponentInteraction)
		{
			if (!interaction.customID.startsWith(CUSTOM_ID_PREFIX))
			{
				return;
			}

			return this.button(interaction);
		}

		if (interaction instanceof CommandInteraction)
		{
			if (interaction.commandName !== SelfRoleCommand.name)
			{
				return;
			}

			if (!interaction.guild)
			{
				await interaction.reply(
					"This guild is not available, the bot might not be fully started yet."
					+ " If this error persists, contact an administrator.",
					{ ephemeral: true },
				);

				return;
			}

			const subCommand = interaction.options.values().next().value as CommandInteractionOption;
			if (!subCommand) throw new Error("No sub command present.");

			const subCommandName = subCommand.name;

			if (isSubCommand(subCommandName))
			{
				this[subCommandName](interaction, subCommand.options!);
			}
		}
	}

	private async button(interaction: MessageComponentInteraction)
	{
		if (interaction.channelID !== channelID) return;
		if (!interaction.guild)
		{
			await interaction.reply(
				"This guild is not available, the bot might not be fully started yet.\n"
				+ "If this error persists, contact an administrator.",
				{ ephemeral: true },
			);

			return;
		}

		const roleID = interaction.customID.slice(CUSTOM_ID_PREFIX.length);

		const role = interaction.guild.roles.resolve(roleID);
		if (!role)
		{
			await interaction.reply(
				"Couldn't find the role for this button.\n"
				+ "If this error, persists contact an administrator.",
				{ ephemeral: true },
			);
			return;
		}

		const selfRole = await getConnection().getRepository(SelfRole).findOne(roleID);
		if (!selfRole)
		{
			await interaction.reply(
				`${role} is not designated to be self assignable.\n`
				+ "If this error, persists contact an administrator.",
				{ ephemeral: true },
			);

			return;
		}

		const member = interaction.member as GuildMember;
		if (member.roles.cache.has(role.id))
		{
			await member.roles.remove(role);
			await interaction.reply(`Successfully removed ${role} from you.`, { ephemeral: true });
		}
		else
		{
			await member.roles.add(role);
			await interaction.reply(`Successfully added ${role} to you.`, { ephemeral: true });
		}
	}

	private async add(interaction: CommandInteraction, options: NonNullable<CommandInteractionOption["options"]>)
	{
		const repo = getConnection().getRepository(SelfRole);

		const roleID = options.get("role")!.role!.id;
		const style = options.get("style")?.value as number ?? MessageButtonStyles.PRIMARY;
		let emoji = options.get("emoji")?.value as string ?? null;

		if (roleID === interaction.guildID)
		{
			await interaction.reply(
				"The everyone role may not be self-assignable; It's the everyone role after all.",
				{ ephemeral: true },
			);
			return;
		}

		// Try to detect custom emojis
		const match = emoji?.match(/^<(a)?:(\w{2,32}):(\d{16,21})>$/);
		if (match)
		{
			emoji = match.slice(1).join("|");
		}

		const selfrole = await repo.findOne(roleID);

		await repo.save({ id: roleID, style, emoji });
		const response = selfrole
			? `The <@&${roleID}> role is now self assignable.\n`
			+ "Make sure to `/selfrole deploy` this change."
			: `Updated style and emoji for the <@&${roleID}> role.\n`
			+ "Make sure to `/selfrole deploy` this change.";

		await interaction.reply(response);
	}

	private async remove(interaction: CommandInteraction, options: NonNullable<CommandInteractionOption["options"]>)
	{
		const repo = getConnection().getRepository(SelfRole);

		const roleID = options.get("role")!.role!.id;
		const selfrole = await repo.findOne(roleID);

		if (selfrole)
		{
			await repo.remove(selfrole);
			await interaction.reply(
				`The <@&${roleID}> role is no longer self assignable.\n`
				+ "Make sure to `/selfrole deploy` this change.",
			);
		}
		else
		{
			await interaction.reply(`The <@&${roleID}> role is not self assignable.`, { ephemeral: true });
		}
	}

	private async show(interaction: CommandInteraction)
	{
		const repo = getConnection().getRepository(SelfRole);

		const roles = await repo.find();

		if (!roles.length)
		{
			await interaction.reply("There are no self assignable roles.");
		}
		else
		{
			const roleString = roles.map(r =>
				`<@&${r.id}> (emoji: ${r.emoji ?? "none"}, style: ${MessageButtonStyles[r.style] ?? r.style})`,
			).join(", ");

			await interaction.reply(
				`Self assignable roles are: ${roleString}`,
			);
		}
	}

	private async deploy(interaction: CommandInteraction, options: NonNullable<CommandInteractionOption["options"]>)
	{
		const repo = getConnection().getRepository(SelfRole);

		const messageID = options.get("message")?.value as Snowflake;

		const channel = this.client.channels.resolve(channelID);
		if (!channel)
		{
			await interaction.reply(
				`The designated channel for self roles (<#${channelID}> / ${channelID}) could not be found.`,
			);

			return;
		}

		if (!(channel instanceof TextChannel))
		{
			await interaction.reply(
				`The designated channel for self roles (<#${channelID}> / ${channelID}) is not a text channel.`,
			);

			return;
		}

		const roles = (await repo.find())
			.map(r => [r, interaction.guild?.roles.resolve(r.id)])
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			.filter(([_, r]) => r) as [SelfRole, Role][];

		const messageOptions: MessageOptions = {
			content: "Click the buttons below to toggle a role on you.",
			components: chunkArray(roles, 5)
				.map<MessageActionRowOptions>(row =>
				({
					type: MessageComponentTypes.ACTION_ROW,
					components: row.map<MessageButtonOptions>(([info, role]) => ({
						type: MessageComponentTypes.BUTTON,
						label: role.name,
						style: info.style,
						customID: `${CUSTOM_ID_PREFIX}${role.id}`,
						emoji: buildEmoji(info.emoji),
					})),
				})),
		};

		try
		{
			if (messageID)
			{
				await channel.messages.edit(messageID, messageOptions);
			}
			else
			{
				await channel.send(messageOptions);
			}
		}
		catch (error: unknown)
		{
			await interaction.reply(
				`Failed to send or edit the message: \`${(error as DiscordAPIError).message}\``,
			);

			throw error;
		}

		const response = messageID
			? `Successfully edited the message for self roles in ${channel}.`
			: `Successfully sent a message for self roles into ${channel}.`
			+ "\nYou may want to delete the old message, if there is one, now.";

		await interaction.reply(response);
	}

	private async cleanup(interaction: CommandInteraction)
	{
		const repo = getConnection().getRepository(SelfRole);

		const selfRoles = await repo.find();
		const missing = selfRoles.filter(r => !interaction.guild!.roles.resolve(r.id));
		if (!missing.length)
		{
			await interaction.reply("There are no roles to prune.", { ephemeral: true });
		}
		else
		{
			await repo.remove(missing);
			await interaction.reply(
				`Removed ${missing.length} no longer existing self assignable roles from the list.`,
			);
		}
	}
}

function isSubCommand(candidate: string): candidate is keyof typeof SubCommand
{
	return Reflect.has(SubCommand, candidate);
}

function buildEmoji(emoji: string | null)
{
	if (!emoji) return undefined;
	if (!emoji.includes("|")) return { name: emoji, id: null };
	const [animated, name, id] = emoji.split("|");

	return {
		animated: animated === "a",
		name,
		id: id as Snowflake,
	};
}

export { SelfRoleModule as Module };

File: src/commands/self_role.rs
use std::collections::HashMap;

use poise::serenity_prelude::{
    model::application::component::ButtonStyle, ChannelId, CreateButton, CreateComponents,
    Mentionable, ReactionType, Role, RoleId,
};

use crate::{models::SelfRole, Context, Error};

#[derive(poise::ChoiceParameter)]
#[repr(i32)]
pub enum StyleOptions {
    #[name = "Primary (Blurple)"]
    Primary = ButtonStyle::Primary as i32,
    #[name = "Secondary (Grey)"]
    Secondary = ButtonStyle::Secondary as i32,
    #[name = "Success (Green)"]
    Success = ButtonStyle::Success as i32,
    #[name = "Danger (Red)"]
    Danger = ButtonStyle::Danger as i32,
}

impl StyleOptions {
    // Allows us to convert a `StyleOptions` value into `ButtonStyle`.
    pub fn convert_from_int(value: i32) -> ButtonStyle {
        match value {
            x if x == StyleOptions::Secondary as i32 => ButtonStyle::Secondary,
            x if x == StyleOptions::Success as i32 => ButtonStyle::Success,
            x if x == StyleOptions::Danger as i32 => ButtonStyle::Danger,
            _ => ButtonStyle::Primary,
        }
    }

    pub fn name_from_int(value: i32) -> &'static str {
        match value {
            x if x == StyleOptions::Primary as i32 => "PRIMARY",
            x if x == StyleOptions::Secondary as i32 => "SECONDARY",
            x if x == StyleOptions::Success as i32 => "SUCCESS",
            x if x == StyleOptions::Danger as i32 => "DANGER",
            _ => "UNKNOWN",
        }
    }
}

/// Creates the common component that we send when deploying self roles.
pub fn create_component<'a>(
    menu: &'a mut CreateComponents,
    roles: HashMap<RoleId, (&Role, SelfRole)>,
) -> &'a mut CreateComponents {
    menu.create_action_row(|row| {
        for role in roles {
            let guild_role = role.1 .0;
            let database_data = role.1 .1;

            let mut button = CreateButton::default();
            button
                .label(&guild_role.name)
                .style(StyleOptions::convert_from_int(database_data.style))
                .custom_id(format!("self-role-{}", database_data.id));

            if let Some(emoji) = database_data.emoji {
                button.emoji(emoji.parse::<ReactionType>().unwrap());
            }

            row.add_button(button);
        }

        row
    })
}

#[poise::command(
    slash_command,
    rename = "selfrole",
    subcommands("add", "remove", "deploy", "show", "cleanup")
)]
pub async fn self_role(_ctx: Context<'_>) -> Result<(), Error> {
    Ok(())
}

// Perhaps we should eventually use selection menus instead of buttons? That way users can only add
// what hasn't already been added.
/// Add a role to the list of available self roles.
#[poise::command(slash_command, ephemeral)]
pub async fn add(
    ctx: Context<'_>,
    #[description = "The role to add to the list of self roles"] role: Role,
    #[description = "The button style to use"] style: StyleOptions,
    #[description = "The emoji to use. (Make sure its actually valid)"] emoji: String,
) -> Result<(), Error> {
    if role.id.as_u64() == ctx.guild().unwrap().id.as_u64() {
        ctx.send(|message| {
            message.content(
                "The everyone role may not be self-assignable; It's the everyone role after all.",
            )
        })
        .await?;

        return Ok(());
    }

    let database_connection = &mut ctx.data().database_connection.get()?;

    // Stores the self role for later checking if there was one already registered or not.
    let selfrole_result = SelfRole::find_one(database_connection, i64::from(role.id));
    SelfRole::upsert(
        database_connection,
        SelfRole {
            id: i64::from(role.id),
            style: style as i32,
            emoji: Some(emoji),
        },
    )?;

    ctx.send(|message| {
        if let Err(_) = selfrole_result {
            message.content(format!(
                concat!(
                    "The <@&{}> role has been added to the list!\n",
                    "Make sure to `/selfrole deploy` this change."
                ),
                role.id
            ))
        } else {
            message.content(format!(
                concat!(
                    "Update style and emoji for the <@&{}> role.\n",
                    "Make sure to `/selfrole deploy` this change."
                ),
                role.id
            ))
        }
    })
    .await?;

    Ok(())
}

/// Remove a role from the list of available self roles.
#[poise::command(slash_command, ephemeral)]
pub async fn remove(
    ctx: Context<'_>,
    #[description = "The role to remove from the list of self roles"] role: Role,
) -> Result<(), Error> {
    let removed = SelfRole::remove(
        &mut ctx.data().database_connection.get()?,
        i64::from(role.id),
    )?;

    let response = if removed {
        format!(
            concat!(
                "The {} role is no longer self assignable!\n",
                "Make sure to `/selfrole deploy` this change."
            ),
            role
        )
    } else {
        format!("The {role} role is not self assignable.")
    };

    ctx.say(response).await?;

    Ok(())
}

/// Deploys the previously configured self roles.
///
/// If you have not configured it yet, do so through `/selfrole add`!
#[poise::command(slash_command, ephemeral)]
pub async fn deploy(
    ctx: Context<'_>,
    #[description = "The id of the message in the self roles channel to update."] message: Option<
        String,
    >,
) -> Result<(), Error> {
    let self_roles = SelfRole::get_all(&mut ctx.data().database_connection.get()?)?;

    let guild = ctx.guild().unwrap();

    // Creates a `HasMap` containing the database's self role data, and the Guild's role data.
    let mut guild_roles = HashMap::new();
    for role in self_roles {
        if let Some(r) = guild.roles.get(&RoleId::from(role.id as u64)) {
            guild_roles.insert(r.id, (r, role));
        }
    }

    let guild_channels = guild.channels(&ctx.serenity_context().http).await?;
    let channel = guild_channels
        .get(&ChannelId::from(
            ctx.data().config.self_roles.channel.parse::<u64>().unwrap(),
        ))
        .unwrap();

    // The menu creation needs to be ported into a macro or function... It's just very annoying to
    // do so, but maybe I'm retarded, it's 5AM after all.
    match message {
        // Checks whether the message provided actually exists.
        Some(msg) => {
            if let Ok(mut m) = channel
                .message(
                    &ctx.serenity_context().http,
                    u64::from(msg.parse::<u64>().unwrap()),
                )
                .await
            {
                m.edit(&ctx.serenity_context().http, |old_message| {
                    old_message
                        .content("Click the buttons below to toggle a role on you.")
                        .components(|menu| create_component(menu, guild_roles))
                })
                .await?;

                ctx.send(|msg| {
                    msg.content(format!(
                        concat!(
                            "Successfully sent a message for self roles into <#{}>",
                            "You may want to delete the old message if there is one, now."
                        ),
                        channel.id
                    ))
                })
                .await?;
            }
        }
        None => {
            channel
                .send_message(&ctx.serenity_context().http, |msg| {
                    msg.content("Click the buttons below to toggle a role on you.")
                        .components(|menu| create_component(menu, guild_roles))
                })
                .await?;

            ctx.send(|msg| {
                msg.content(format!(
                    concat!(
                        "Successfully edited the message for self roles in <#{}>",
                        "You may want to delete the old message if there is one, now."
                    ),
                    channel.id
                ))
            })
            .await?;
        }
    }

    Ok(())
}

/// Show all available self roles.
#[poise::command(slash_command, ephemeral)]
pub async fn show(ctx: Context<'_>) -> Result<(), Error> {
    let roles = SelfRole::get_all(&mut ctx.data().database_connection.get()?)?
        .iter()
        .map(|self_role| {
            format!(
                "{} (emoji: {}, style: {})",
                RoleId::from(self_role.id as u64).mention(),
                self_role.emoji.as_ref().unwrap_or(&String::from("none")),
                StyleOptions::name_from_int(self_role.style)
            )
        })
        .collect::<Vec<String>>()
        .join(", ");

    let response = if roles.is_empty() {
        "There are not self assignable roles.".into()
    } else {
        format!("Self assignable roles are: {}", roles)
    };

    ctx.say(response).await?;

    Ok(())
}

/// Removes deleted roles from the list of self assignable roles.
#[poise::command(slash_command, ephemeral)]
pub async fn cleanup(ctx: Context<'_>) -> Result<(), Error> {
    let connection = &mut ctx.data().database_connection.get()?;

    let guild_roles = ctx.guild().unwrap().roles;

    let deleted_roles: Vec<i64> = SelfRole::get_all(connection)?
        .iter()
        .filter_map(|self_role| {
            if guild_roles.contains_key(&RoleId::from(self_role.id as u64)) {
                Some(self_role.id)
            } else {
                None
            }
        })
        .collect();

    let response = if deleted_roles.is_empty() {
        "There are no roles to prune.".into()
    } else {
        SelfRole::remove_all(connection, &deleted_roles)?;

        format!(
            "Removed {} no longer existing self assignable roles from the list.",
            deleted_roles.len()
        )
    };

    ctx.say(response).await?;

    Ok(())
}
